File: src/main/scala/fi/oph/koski/schema/OsaamisenTunnustaminen.scala
package fi.oph.koski.schema

import fi.oph.koski.schema.annotation._
import fi.oph.scalaschema.annotation.{DefaultValue, Description}

@Description("Tiedot aiemmin hankitun osaamisen tunnustamisesta.")
@OksaUri("tmpOKSAID629", "osaamisen tunnustaminen")
case class OsaamisenTunnustaminen(
  @Description("Aiemman, korvaavan suorituksen tiedot")
  @FlattenInUI
  osaaminen: Option[Suoritus],
  selite: LocalizedString,
  @Description("Käytetään, mikäli tunnustettu osaaminen kuuluu rahoitukseen piiriin (esimerkiksi kaksoistutkintolaisilla ammatilliseen tutkintoon tunnustetut lukio-opinnot tai toiselta oppilaitokselta ostetut yksittäiset tutkinnon osat).")
  @Tooltip("Tunnustettu osaaminen kuuluu rahoitukseen piiriin (esimerkiksi kaksoistutkintolaisilla ammatilliseen tutkintoon tunnustetut lukio-opinnot tai toiselta oppilaitokselta ostetut yksittäiset tutkinnon osat).")
  @DefaultValue(false)
  rahoituksenPiirissä: Boolean = false
) extends SelitettyOsaamisenTunnustaminen

trait SelitettyOsaamisenTunnustaminen {
  @Description("Osaamisen tunnustamisen kautta saatavan tutkinnon osan suorituksen selite.")
  @Tooltip("Kuvaus siitä, miten aikaisemmin hankittu osaaminen on tunnustettu.")
  @OksaUri("tmpOKSAID629", "osaamisen tunnustaminen")
  @Representative
  @MultiLineString(5)
  def selite: LocalizedString
}

trait MahdollisestiTunnustettu {
  def tunnustettu: Option[SelitettyOsaamisenTunnustaminen]

  def tunnustettuRahoituksenPiirissä: Boolean = tunnustettu.exists {
    case o: OsaamisenTunnustaminen => o.rahoituksenPiirissä
    case _ => false
  }
}

File: web/app/types/fi/oph/koski/schema/OsaamisenTunnustaminen.ts
import { Suoritus } from './Suoritus'
import { LocalizedString } from './LocalizedString'

/**
 * Tiedot aiemmin hankitun osaamisen tunnustamisesta.
 *
 * @see `fi.oph.koski.schema.OsaamisenTunnustaminen`
 */
export type OsaamisenTunnustaminen = {
  $class: 'fi.oph.koski.schema.OsaamisenTunnustaminen'
  osaaminen?: Suoritus
  selite: LocalizedString
  rahoituksenPiirissä: boolean
}

export const OsaamisenTunnustaminen = (o: {
  osaaminen?: Suoritus
  selite: LocalizedString
  rahoituksenPiirissä?: boolean
}): OsaamisenTunnustaminen => ({
  $class: 'fi.oph.koski.schema.OsaamisenTunnustaminen',
  rahoituksenPiirissä: false,
  ...o
})

OsaamisenTunnustaminen.className =
  'fi.oph.koski.schema.OsaamisenTunnustaminen' as const

export const isOsaamisenTunnustaminen = (a: any): a is OsaamisenTunnustaminen =>
  a?.$class === 'fi.oph.koski.schema.OsaamisenTunnustaminen'
